SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/RemotePossessionComponent.cpp

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/Components/RemotePossessionComponent.h"
+#include "LoadBalancing/AbstractLBStrategy.h"
+
+#include "Engine/World.h"
+
+DEFINE_LOG_CATEGORY(LogRemotePossessionComponent);
+
+URemotePossessionComponent::URemotePossessionComponent(const FObjectInitializer& ObjectInitializer)
+	: Super(ObjectInitializer)
+	, PendingDestroy(false)
+{
+	PrimaryComponentTick.bCanEverTick = true;
+	PrimaryComponentTick.bStartWithTickEnabled = true;
+
+#if ENGINE_MINOR_VERSION <= 23
+	bReplicates = true;
+#else
+	SetIsReplicatedByDefault(true);
+#endif
+}
+
+void URemotePossessionComponent::OnAuthorityGained()
+{
+	AController* Controller = Cast<AController>(GetOwner());
+	ensure(Controller);
+	if (Target == nullptr)
+	{
+		OnInvalidTarget();
+		return;
+	}
+	if (!Target->HasAuthority())
+	{
+		UE_LOG(LogRemotePossessionComponent, Verbose, TEXT("Worker is not authoritative over target: %s"), *Target->GetName());
+		return;
+	}
+	else
+	{
+		if (EvaluatePossess())
+		{
+			UE_LOG(LogRemotePossessionComponent, Verbose, TEXT("Remote possession succesful on (%s)"), *Target->GetName());
+			Controller->Possess(Target);
+		}
+		else
+		{
+			UE_LOG(LogRemotePossessionComponent, Verbose, TEXT("EvaluatePossess(%s) failed"), *Target->GetName());
+		}
+		MarkToDestroy();
```

Comment:
What's the reasoning behind marking for destruction, instead of calling DestroyComponent directly?

Reply:
If call DestroyComponent directly, it will be crashed at OnAuthorityGained() because of iterator invalid.
DestroyComponent changed the OwnedComponents from MyOwner->RemoveOwnedComponent(this);
Actor::OnAuthorityGained for each OwnedComponents and call OnAuthorityGained() of them. It will be broken on for each.

Reply:
Ah I believe this is fixed now as of https://github.com/spatialos/UnrealGDK/pull/2800